// knub/dist/CooldownManager.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const CLEANUP_INTERVAL = 1000 * 60 * 5; // 5min
class CooldownManager {
    constructor() {
        this.cooldowns = new Map();
        setTimeout(() => this.cleanup(), CLEANUP_INTERVAL);
    }
    cleanup() {
        const now = Date.now();
        for (const [key, cdEnd] of this.cooldowns.entries()) {
            if (cdEnd < now)
                this.cooldowns.delete(key);
        }
        setTimeout(() => this.cleanup(), CLEANUP_INTERVAL);
    }
    setCooldown(key, time) {
        const cdEnd = Date.now() + time;
        this.cooldowns.set(key, cdEnd);
    }
    isOnCooldown(key) {
        if (!this.cooldowns.has(key))
            return false;
        return this.cooldowns.get(key) >= Date.now();
    }
    getCooldownRemaining(key) {
        if (!this.isOnCooldown(key))
            return 0;
        return Math.max(0, Date.now() - this.cooldowns.get(key));
    }
}
exports.CooldownManager = CooldownManager;
